src/site/apt/matcher_def.apt.vm

```diff
@@ -0,0 +1,364 @@
+@FunctionalInterface
+interface Builder {
+    /**
+     * Build the IHeaderMatcher.
+     * 
+     * Implementations of this interface should return a specific child class of IHeaderMatcher.
+     * 
+     * @return a new IHeaderMatcher.
+     */
+    IHeaderMatcher build();
+
+    /**
+     * Gets the class that is build by this builder.
+     * @return The class that is build by this builder.
+     */
+    default Class<?> builtClass() throws SecurityException {
+        try {
+            return this.getClass().getMethod("build").getReturnType();
+        } catch (NoSuchMethodException | SecurityException e) {
+            throw new IllegalStateException("the 'build' method of the Builder interface must alwasy be public");
+        }
+    }
+
+    /**
+     * Gets the Description for this builder.
+     * @return The description of the builder 
+     */
+    default Description getDescription() {
+        Class<?> clazz = builtClass();
+        if (clazz == IHeaderMatcher.class) {
+            throw new ImplementationException(String.format(
+                    "Class %s must implement built() method to return a child class of IHeaderMatcher", 
+                    this.getClass()));
+        }
+        return DescriptionBuilder.buildMap(clazz);
+    }
+}
++------------------------------------------+
+
+  The work of handling the Id and some other tasks is handled by the AbstractBuilder.
+  
++------------------------------------------+
+public abstract class AbstractBuilder implements IHeaderMatcher.Builder {
+
+    private String id;
+
+    /**
+     * Protected empty constructor.
+     */
+    protected AbstractBuilder() {
+    }
+
+    /**
+     * Set the id for the matcher.
+     * @param id the id to use.
+     * @return this builder for chaining.
+     */
+    public final AbstractBuilder setId(String id) {
+        this.id = id;
+        return this;
+    }
+
+    /**
+     * @return {@code true} if the id is not null and not blank.
+     */
+    public final boolean hasId() {
+        return !StringUtils.isBlank(id);
+    }
+
+    /**
+     * @return the id as specified in the builder.
+     */
+    protected String getId() {
+        return id;
+    }
+
+    @Override
+    public String toString() {
+        return String.format("%s with id %s", this.getClass(), id);
+    }
+}  
++------------------------------------------+
+  
+  So we have:
+
++------------------------------------------+
+package com.example.ratMatcher;
+
+import org.apache.rat.configuration.builders.AbstractBuilder;
+
+public class QSCBuilder extends AbstractBuilder. {
+    QSCMatcher build() {
+        return new QSCMatcher(getId());
+    }
+}
++------------------------------------------+
+
+** Registering the builder for use in XML configuration 
+
+  In order to use the matcher in a Rat configuration it has to be registered with the system.  This can be done by creating file with the builder specified and passing it
```

Comment:
this can be done by creating a file? What kind of file? XML or configuration file?

Reply:
Maybe the type just needs to be added for clarification purposes.